server/src/main/java/org/elasticsearch/plugins/SearchPlugin.java

```diff
@@ -124,6 +125,16 @@
         return emptyList();
     }
 
+    /**
+     * Allows plugins to register cache differentiators which contributes to the cacheKey
+     * computation for the request cache. This helps differentiate between queries that
+     * are otherwise identical.
+     * @return
+     */
+    default ShardSearchRequest.RequestCacheKeyProvider getRequestCacheKeyProvider() {
```

Comment:
This was a `BiConsumer` but got deliberately changed to a named class as a result of simplifying/tightening this extension point. Please let me explain: 

A Provider returned from this method is now used exclusively for computing `cacheKey` of a `ShardSearchRequest`. That is, the Provider is no longer called inside `ShardSearchRequest#cacheKey` method after existing code is executed. The relationship is flipped, i.e. the Provider is now the caller and has the responsibility to ensure existing code of cacheKey always runs. So instead of
```java
request.cacheKey(provider);
```
We now have
```java
provider.cacheKey(request);
```
We feel the new form reads more natural because it feels somewhat awkward to expose this additional cacheKey logic of provider to ShardSearchRequest. From the perspective of `IndicesService` (caller of `ShardSearchRequest#cacheKey`), it seems more fluent to say, "give me a provider for computing cacheKey and I'll pass any `ShardSearchRequest` to it".

With above being said, a concern is that any Provider implementations now have the responsibility to ensure the existing logic in `ShardSearchRequest#cacheKey` always gets executed. To enforce this, we chose to have a named base class with the mandatory logic being encapsulated in a public final method (`cacheKey`) and a protected method (`differentiate`) for subclass (`Security` in this case) to overide for additional logic.

Reply:
Ok, maybe the confusion comes from the name, `RequestCacheKeyProvider` sounds like it's the responsibility of the plugin to generate a cache key. The real intent here is to allow a plugin to append custom bytes to the existing cache key. We could for instance change `ShardSearchRequest#cacheKey` to take the extra consumer into account:
```
public BytesReference cacheKey(BiConsumer<ShardSearchRequest, StreamOutput> extraCacheKeyConsumer) throws IOException {
```
We have the same mechanism to wrap the directory reader and we don't need a new class to enforce some logic there.

Reply:
The method signature you suggested is in fact the previous version I had. After discussion with Tim, we agreed to change it to the current approach. 

> RequestCacheKeyProvider sounds like it's the responsibility of the plugin to generate a cache key

You are reading it correctly. The curret approach does move the responsibility to the plugin. To ensure/help the Plugin fulfill its responsibility, we went with the named based class. So we went from "plugin needs to append some bytes" to "plugin provides the whole cache key and it also ensures existing bytes are maintained".

Since this is search code, I am happy to revert the changes and stick to "Plugin needs to append some bytes with a BiConsumer". @tvernum What do you think?

Reply:
I'm OK with any option.

My personal view is that it's OK for the IndicesService to just delegate to a `Request -> BytesReference` function and trust that plugins will do the right thing.
But I'm OK with a `(Request, StreamOutput)` consumer.

Reply:
Thanks Tim. I updated the PR to re-take the previous approach of `CheckedBiConsumer` which is used as an extension for existing cache key calculation. Also made sure only a single plugin can register it.